Openemr fix 4959 amc pi hie 1

adunsulag opened the pull request:
Implement the Send Referral loops code.  Starting this off as a draft so people can see what I'm doing. 

Created and implemented the AMC 315g Required Rule 7 which is the Support Electronic Referral Loops by Sending Health Information rule for MIPS.

Updated the transactions help to discuss MIPS reporting pieces.  Implemented a help file for the AMC report that also explains how to run the report and specific information for MIPS calculations.

Added a new checkbox flag on the transactions for providers to confirm
that the recipient they sent a summary of care document to actually
received the document.

Made it so the confirmation receipt is set to true when EMRDirect is sent since EMRDirect automatically confirms that the recipient received the message.

Added facility selector to referral form.  Changed up the AMC report to use the billing_facility_id.

Still need to update the ccda amc_sum change to make sure that is working but I can at least now generate the baseline report. Will do more tests on ONC and make sure that is all working properly.

stephenwaite commented on sql/6_1_0-to-7_0_0_upgrade.sql:
```diff
@@ -211,9 +211,27 @@ INSERT INTO `list_options` (`list_id`, `option_id`, `title`, `seq`, `activity`)
 INSERT INTO `layout_options` (`form_id`,`field_id`,`group_id`,`title`,`seq`,`data_type`,`uor`,`fld_length`,`max_length`
-    ,`list_id`,`titlecols`,`datacols`,`default_value`,`edit_options`,`description`,`fld_rows`)
-    VALUES ('LBTref', 'encounter_id', '1', 'Patient Referral Encounter', 10, 53, 1, 0, 0, '', 1, 1, '', ''
-    ,'Encounter that the referral/transfer of care is based on', 0);
```
fare thee well poor encounter_id :crying_cat_face:

bradymiller commented on sql/6_1_0-to-7_0_0_upgrade.sql:
```diff
@@ -238,3 +231,37 @@ INSERT INTO `list_options` (`list_id`, `option_id`, `title`, `seq`, `is_default`
+       , '170.315(g)(1)/(2)–7', '0', '0', '1', '0', '', '', '', '', '', '', '');
+#EndIf
+
+#IfNotRow2D layout_options form_id LBTref field_id billing_facility_id
```
would drop the encounter_id stuff from database here

adunsulag replied:
Added in the drop fix.

bradymiller commented on sql/6_1_0-to-7_0_0_upgrade.sql:
```diff
@@ -238,3 +231,37 @@ INSERT INTO `list_options` (`list_id`, `option_id`, `title`, `seq`, `is_default`
+#EndIf
+
+#IfMissingColumn report_itemized item_details
+ALTER TABLE `report_itemized` ADD COLUMN `item_details` TEXT DEFAULT NULL;
```
keep TEXT bare

adunsulag replied:
Resolved.

bradymiller commented on sql/database.sql:
```diff
@@ -1064,8 +1064,12 @@ INSERT INTO `clinical_rules`
+
+-- 2015 AMC Rules
+INSERT INTO `clinical_rules` (`id`, `pid`, `active_alert_flag`, `passive_alert_flag`, `cqm_flag`, `cqm_2011_flag`,`cqm_2014_flag`, `cqm_nqf_code`, `cqm_pqri_code`, `amc_flag`, `amc_2011_flag`,`amc_2014_flag`, `amc_code`, `amc_code_2014`, `amc_code_2015`, `amc_2014_stage1_flag`,`amc_2014_stage2_flag`, `amc_2015_flag`, `patient_reminder_flag`, `developer`,`funding_source`, `release_version`, `web_r ...
+VALUES ('patient_access_amc', '0', '0', '0', '0', '0', '0', '', '', '1', '0', '0', '', '', '170.315(g)(1)/(2)–2c', '0', '0', '1', '0', '', '', '', '', '', '', '');
```
why was `patients:med` removed from this?

adunsulag replied:
I saw that the other AMC rules didn't have the patients:med reqirement so I removed it as well.  I can put it back if needed.

bradymiller commented on sql/database.sql:
```diff
@@ -7706,6 +7712,8 @@ CREATE TABLE `report_itemized` (
   `pass` tinyint(1) NOT NULL DEFAULT '0' COMMENT '0 is fail, 1 is pass, 2 is excluded',
   `pid` bigint(20) NOT NULL,
+  `rule_id` VARCHAR(31) DEFAULT NULL  COMMENT 'fk to clinical_rules.rule_id',
+  `item_details` TEXT DEFAULT NULL COMMENT 'JSON with specific sub item results for a clinical rule',
```
make text bare

adunsulag replied:
Resolved this issue.

bradymiller commented on templates/reports/cqm/_results-table-patient-details.html.twig:
```diff
@@ -0,0 +1,80 @@
+                            <tr>
+                                <td class="amc-demographic-label">{{ patientDetails.patient.fname }} {{ patientDetails.patient.lname}}</td>
+                                <td class="amc-demographic-label">{{ patientDetails.patient.sex }}</td>
+                                <td class="amc-demographic-label">{{ patientDetails.patient.DOB }}</td>
```
above 3 lines have not been html escaped

adunsulag replied:
Fixed this now.

bradymiller commented on templates/reports/cqm/_results-table-patient-details.html.twig:
```diff
@@ -0,0 +1,80 @@
+                                    {%  for action in patientDetails.actions %}
+                                    <td class="amc-action-label">
+                                        <strong>{% if action.value %}{{ "Yes"|xlt }}{% else %}{{ "No"|xlt }}{% endif %}</strong>
+                                        {% if action.details != "" %} - <span>{{ action.details }}</span>{% endif %}
```
action.details not escaped (possible that you have done this already though)

adunsulag replied:
Should be fixed now.

bradymiller commented on sql/6_1_0-to-7_0_0_upgrade.sql:
```diff
@@ -238,3 +231,42 @@ INSERT INTO `list_options` (`list_id`, `option_id`, `title`, `seq`, `is_default`
+#EndIf
+
+#IfMissingColumn ccda transaction_id
+ALTER TABLE `ccda` ADD COLUMN `transaction_id` BIGINT(20) COMMENT 'fk to transactions referral record';
```
this new column also needs to be added in database.sql

adunsulag replied:
Added this in.

bradymiller commented:
@adunsulag , code looks great! just need to add the transaction_id column to database.sql and then ready for codebase

adunsulag commented:
@bradymiller Hey Brady, I thought I was done here, but had to make a number of changes in 11 files.  If you can do another pass through this when you get a chance I'll get them resolved.  Course my fingers are crossed it all looks good and can just be merged :)

bradymiller commented:
@adunsulag , looks great!

![2DV-2](https://user-images.githubusercontent.com/278968/161415288-4d3758ba-3bba-44ce-9525-d01773eb445f.gif)

adunsulag commented:
@bradymiller Thought I was done this morning, but this is the last bit haha!  I went ahead and cleaned up the cqm report to remove unused fields for the current AMC rule set.  I also decided we just don't to execute the old AMC rulesets as there's no point for MIPs reporting.  I made it so all of the old reports still load but that for new reports you get the limited subset of AMC selections.  

The standard alert report selections were cleaned up to remove AMC features. 

I revamped the patient details report as the printing was going completely haywire.  It can still be a bit finicky due to the table columns but I think I got things resolved.  

I also added in batching so that the MySQL server doesn't die.

At this point all AMC scenarios pass and when this PR is brought in AMC will be code complete.

bradymiller commented on templates/reports/cqm/cqm-form.html.twig:
```diff
@@ -0,0 +1,153 @@
+                                    ({{ 'Non-electronic'|xlt }})
+                                </td>
+                                <td>
+                                    <input {{ dis_text|text }} type="text" id="labs_manual_entry" name="labs_manual_entry" class='form-control' value="{{ labs_manual|attr }}">
```
will escaping dis_text break this? (looks like it has quotes in it that are likely needed for html and is not escaped above)

adunsulag replied:
As far as I can tell escaping it didn't break it.

bradymiller commented:
looks good to go! amc is done!
![2DV-2](https://user-images.githubusercontent.com/278968/161681578-7848a098-6117-4565-affa-5a81fea6ad86.gif)
